# Drawing Conclusions Using Groupby

Use `winequality_edited.csv`. You should've created this data file in the previous section: *Appending Data (cont.)*.

In [2]:
# import panda
import pandas as pd

# Load dataset

df = pd.read_csv('winequality_edited.csv')


### Is a certain type of wine associated with higher quality?

In [3]:
# Example Usage of groupby()

groupby_mean_quality_table = df.groupby('quality').mean()

groupby_mean_quality_color_table = df.groupby(['quality','color']).mean()

pH_mean_groupby_quality_color_no_idx = df.groupby(
                        ['quality','color'], as_index=False)['pH'].mean()

## Q1: Is a certain type of wine (red or white) associated with higher quality?
For this question, compare the average quality of red wine with the average quality of white wine with groupby. To do this group by color and then find the mean quality of each group.

In [4]:
# Find the mean quality of each wine type (red and white) with groupby
df.groupby(['color'])['quality'].mean()

color
red      5.636023
white    5.877909
Name: quality, dtype: float64

## Results:
From this dataset, **white wine** received an average quality rating of **5.88**, while **red wine** with average quality rating of **5.64**.

**white wine** is rated **0.24 or a quarter of a point** higher than red wine.

### What level of acidity receives the highest average rating?

In [5]:
# View the min, 25%, 50%, 75%, max pH values with Pandas describe
df.describe().pH

count    6497.000000
mean        3.218501
std         0.160787
min         2.720000
25%         3.110000
50%         3.210000
75%         3.320000
max         4.010000
Name: pH, dtype: float64

In [6]:
# Bin edges that will be used to "cut" the data into groups
bin_edges = [2.72, 3.11, 3.21, 3.32, 4.01] # Fill in this list with five values you just found

In [13]:
# Labels for the four acidity level groups
bin_names = ["high", "mod_high", "med", "low"] # Name each acidity level category

In [14]:
# Creates acidity_levels column
df['acidity_levels'] = pd.cut(df['pH'], bin_edges, labels=bin_names)

# Checks for successful creation of this column
df.head()

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality color acidity_levels  
0      9.4        5   red            low  
1      9.8        5   red       mod_high  
2      9.8        5   red            med  
3      9.8        6   red       mod_high  
4      9.4        5   red            low

In [15]:
# Find the mean quality of each acidity level with groupby
df.groupby(['acidity_levels']).quality.mean()

acidity_levels
high        5.783343
mod_high    5.784540
med         5.850832
low         5.859593
Name: quality, dtype: float64

### Results:

Wines with the **lowest acidity** has **highest quality rating** 

In [16]:
# Save changes for the next section
df.to_csv('winequality_edited.csv', index=False)